### PR TITLE
Add sndio module.

### PR DESCRIPTION
**Pull request by ericonr**
- can control sndio: change volume, toggle mute
- appearance is somewhat dynamic: takes muted status into account
- uses polling inside sleeper thread to update values

### CODE REVIEW

**Comment by ericonr** [comment]
Hmm I believe this will also require a symbol check, since I'm using functions introduced in version 1.7.0 of sndio.

File: src/modules/sndio.cpp [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    int revents = sioctl_revents(hdl_, pfds_.data());
+    if (revents & POLLHUP) {
+      throw std::runtime_error("disconnected!");
+    }
+  };
+}
+
+Sndio::~Sndio() {
+  sioctl_close(hdl_);
+}
```
Comment (ericonr) [inline src/modules/sndio.cpp:75]: Does this destructor also kill the thread, or would I need to do it explicitly?

Comment (Alexays) [reply src/modules/sndio.cpp:75]: Nope should be good :)

**Comment by ericonr** [comment]
@Alexays did this compile for FreeBSD with `sndio` support? I can't see CI results. I have now added `sndio-dev` to the alpine dockerfile as well, to increase the coverage.

File: src/modules/sndio.cpp [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    }
+    while (poll(pfds_.data(), nfds, -1) < 0) {
+      if (errno != EINTR) {
+        throw std::runtime_error("poll() failed.");
+      }
+    }
+
+    int revents = sioctl_revents(hdl_, pfds_.data());
+    if (revents & POLLHUP) {
+      throw std::runtime_error("disconnected!");
```
Comment (ericonr) [inline src/modules/sndio.cpp:68]: Due to this, restarting `sndiod` (which should be rare) makes Waybar abort. How would I throw an error that indicates to `waybar` that this module should be turned off?

```
terminate called after throwing an instance of 'std::runtime_error'
  what():  disconnected!
fish: 'waybar' terminated by signal SIGABRT (Abort)
```

Comment (ericonr) [reply src/modules/sndio.cpp:68]: It could also try reconnecting...

File: .travis.yml [hunk 3]
```diff
@@ -29,7 +29,7 @@ jobs:
       compiler: clang
       env:
       before_install:
-        - sudo pkg install -y gtk-layer-shell gtkmm30 jsoncpp libdbusmenu
+        - sudo pkg install -y gtk-layer-shell gtkmm30 jsoncpp libdbusmenu sndio
```
Comment (jbeich) [inline .travis.yml:32]:
```suggestion
        - export CPPFLAGS=-isystem/usr/local/include LDFLAGS=-L/usr/local/lib # sndio
        - sudo pkg install -y gtk-layer-shell gtkmm30 jsoncpp libdbusmenu sndio
```

**Review by jbeich — COMMENTED** [review]
> @Alexays did this compile for FreeBSD with sndio support? I can't see CI results.

Looking at https://travis-ci.org/github/Alexays/Waybar/pull_requests

> Library sndio found: NO

Adding `CPPFLAGS=-isystem/usr/local/include` + `LDFLAGS=-L/usr/local/lib` leads to

> Library sndio found: YES
> Checking for function "sioctl_open" with dependency -lsndio: YES

Seems to build fine after that.

File: .travis.yml (hunk 3 above)
Comment (ericonr) [reply .travis.yml:32]: Why not use `-I`?

Thanks for testing and for the link to CI results! Do you use `sndio`? Would be nice to get opinions on usability and such.

Comment (jbeich) [reply .travis.yml:32]: > Why not use `-I`?

`-isystem` puts include path after `-I` (avoiding conflicts with local headers) and silences warnings (that cannot be fixed by this project).

See also https://gcc.gnu.org/onlinedocs/cpp/System-Headers.html

> Do you use `sndio`? Would be nice to get opinions on usability and such.

No. Maybe @t6 can help.

Comment (ericonr) [reply .travis.yml:32]: I see, thanks!

**Comment by jbeich** [comment]
From https://travis-ci.org/github/Alexays/Waybar/jobs/727586703
> [88/97] Installing sndio-1.6.0...
[...]
> Library sndio found: YES
> Checking for function "sioctl_open" with dependency -lsndio: NO 

Until 2020-10-10 you may need to revert #772 (e.g., [commit](https://github.com/Alexays/Waybar/commit/81ae5ba3a85939cd62aec4e76b47cee456e51b8d), [build](https://travis-ci.com/github/jbeich/Waybar/jobs/385412638)). https://github.com/freebsd/freebsd-ports/commit/68eb5c3a71f59793b00a22f9390d84e4375c8c7f is missing on 2020Q3 but 2020Q4 will be branched in ~2 weeks + ~5 days to build due to many updates.

**Comment by ericonr** [comment]
I see, thanks. I guess I will revert to get it to build at least once in CI, then I can revert it.

**Comment by ericonr** [comment]
@jbeich thanks a lot, it did build with `sndio` support now: https://travis-ci.org/github/Alexays/Waybar/jobs/727695954

I think now that a job has run I should move back to `/quarterly`, right?

**Comment by ericonr** [comment]
Rebased on latest.

File: src/modules/sndio.cpp (hunk 2 above)
Comment (Alexays) [reply src/modules/sndio.cpp:68]: Go for implementing reconnecting if it's not too complex.

**Comment by Alexays** [comment]
Beside the reconnection part, it looks perfect!
Waiting for the first point.
Thanks!

File: src/modules/sndio.cpp (hunk 2 above)
Comment (ericonr) [reply src/modules/sndio.cpp:68]: I will have to look into how to do that without blocking. Will let you know how it goes.

**Comment by ericonr** [comment]
@Alexays thanks for the incentive, reconnection now works! It should be good to go, then.

File: src/modules/sndio.cpp [hunk 4]
```diff
@@ -20,27 +21,28 @@ void onval(void *arg, unsigned int addr, unsigned int val) {
...
   hdl_ = sioctl_open(SIO_DEVANY, SIOCTL_READ | SIOCTL_WRITE, 0);
   if (hdl_ == nullptr) {
     throw std::runtime_error("sioctl_open() failed.");
   }
 
-  if(sioctl_ondesc(hdl_, ondesc, this) == 0) {
-    throw std::runtime_error("sioctl_ondesc() failed.");
-  }
-
+  sioctl_ondesc(hdl_, ondesc, this);
```
Comment (Alexays) [inline src/modules/sndio.cpp:30]: Shouldn't we keep the check here?

File: src/modules/sndio.cpp [hunk 5]
```diff
@@ -65,7 +67,23 @@ Sndio::Sndio(const std::string &id, const Json::Value &config)
 
     int revents = sioctl_revents(hdl_, pfds_.data());
     if (revents & POLLHUP) {
-      throw std::runtime_error("disconnected!");
+      spdlog::warn("sndio disconnected!");
+      sioctl_close(hdl_);
+      hdl_ = nullptr;
+
+      // reconnection loop
+      while (1) {
```
Comment (Alexays) [inline src/modules/sndio.cpp:75]: Could you replace with `while (thread_.isRunning())`?

**Comment by Alexays** [comment]
Two small comments and then it's good.

File: src/modules/sndio.cpp (hunk 4 above)
Comment (ericonr) [reply src/modules/sndio.cpp]: I was a bit unclear on this function's failure modes, but a better look at the source confirmed that we should indeed check for errors. I added a check for `sioctl_onval` too.

File: src/modules/sndio.cpp (hunk 5 above)
Comment (ericonr) [reply src/modules/sndio.cpp]: Just did.

File: src/modules/sndio.cpp [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+      sioctl_close(hdl_);
+      hdl_ = nullptr;
+
+      // reconnection loop
+      while (thread_.isRunning()) {
+        try {
+          connect_to_sndio();
+        } catch(std::runtime_error const& e) {
+          // avoid leaking hdl_
+          if (hdl_) sioctl_close(hdl_);
```
Comment (ericonr) [inline src/modules/sndio.cpp:85]: Added this.

**Comment by Alexays** [comment]
Thanks!
Can you update the github wiki?

**Comment by ericonr** [comment]
Sure! Thanks :D

**Pull request merged**